stories/spectra/mass.stories.tsx

```diff
@@ -72,62 +84,120 @@ export function AdvancedMassExample({ mf }: AdvancedMassExampleProps) {
-        onMouseDown={({ coordinates: { x } }) => {
-          setPositions({
-            position: {
-              x1: x,
-              x2: x,
-            },
-            min: min,
-            max: max,
-          });
+        svgStyle={{
+          cursor: `${alt ? (click.current ? 'grabbing' : 'grab') : ''}`,
+        }}
+        onMouseDown={({ coordinates: { x, y } }) => {
+          if (alt) {
+            setPositions((positions) => ({
+              ...positions,
+              hand: {
+                x,
+                y,
+              },
+            }));
+          } else {
+            setPositions((positions) => ({
+              ...positions,
+              rectangle: {
+                x1: x,
+                x2: x,
+              },
+            }));
+          }
           click.current = true;
         }}
+        onKeyDown={({
+          event: { altKey },
+          x: { max: maxX, min: minX },
+          y: { max: maxY, min: minY },
+        }) => {
+          if (!click.current) {
+            setPositions((positions) => ({
+              ...positions,
+              alt: altKey,
+              maxX,
+              minX,
+              maxY,
+              minY,
+            }));
+          }
+        }}
+        onKeyUp={({ event: { altKey } }) => {
+          if (!click.current || alt) {
+            click.current = false;
+            setPositions((positions) => ({ ...positions, alt: altKey }));
+          }
+        }}
         onMouseUp={() => {
-          click.current = false;
-          if (position.x1 !== position.x2) {
-            setPositions({
-              position: null,
-              min: Math.min(position.x1, position.x2),
-              max: Math.max(position.x1, position.x2),
-            });
+          if (click.current && !alt && rectangle.x1 !== rectangle.x2) {
+            setPositions((positions) => ({
+              ...positions,
+              rectangle: null,
+              minX: Math.min(rectangle.x1, rectangle.x2),
+              maxX: Math.max(rectangle.x1, rectangle.x2),
+            }));
           }
+          click.current = false;
         }}
         onMouseLeave={() => {
-          setPositions({
-            position: null,
-            min: min,
-            max: max,
-          });
+          setPositions((positions) => ({
+            ...positions,
+            rectangle: null,
+          }));
           click.current = false;
         }}
-        onMouseMove={({ coordinates: { x } }) => {
+        onMouseMove={({ coordinates: { x, y } }) => {
           if (click.current) {
-            setPositions(({ position }) => ({
-              position: {
-                x1: position ? position.x1 : x,
-                x2: x,
-              },
-              min: min,
-              max: max,
-            }));
+            if (alt) {
+              setPositions((positions) => ({
+                ...positions,
+                maxX: hand.x - x + maxX,
```

Comment:
You should probably use https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/movementX and don't keep any previous position in variables

Reply:
fixed